File: map_py/gui/game_state_listener.py
import socketio
import threading
from map_py.observations_and_actions.pydantic_obs import format_pydantic_observation
import map_py.gui.visualizer as visualizer
from collections import defaultdict

from collections import deque

class Queue:
    def __init__(self):
        self._items = deque()

    def enqueue(self, item):
        """Add an item to the back of the queue."""
        self._items.append(item)

    def dequeue(self):
        """Remove and return the item at the front of the queue."""
        if self.is_empty():
            raise IndexError("dequeue from an empty queue")
        return self._items.popleft()

    def peek(self):
        """Return the item at the front without removing it."""
        if self.is_empty():
            return {}
        return self._items[0]

    def is_empty(self):
        """Check if the queue is empty."""
        return len(self._items) == 0

    def size(self):
        """Return the number of items in the queue."""
        return len(self._items)
    
    def clear(self):
        """Remove all items from the queue."""
        self._items = deque()

    def __repr__(self):
        return f"Queue({list(self._items)})"


class GameStateListener:
    def __init__(self, buffer: Queue, accept_midday_updates: bool = False):
        self.sio = socketio.Client()
        self.accept_midday_updates = accept_midday_updates
        
        # Step and buffer management
        self.buffer = buffer
        self.num_updates = 0

    def start_socketio_listener(self, server="http://localhost:3000"):
        def run():
            try:
                self.sio.connect(server)
                self.sio.wait()
            except Exception as e:
                print(f"[SocketIO] Connection failed: {e}")

        @self.sio.event
        def connect():
            print("[SocketIO] Connected to server")

        @self.sio.event
        def disconnect():
            print("[SocketIO] Disconnected from server")

        @self.sio.on("action")
        def on_action(data):
            print(data)

        @self.sio.on("game_update")
        def on_update(data):
            if data["state_type"] in ["full_state", "day_start", "day_end"]:
                staff_list = data["data"]["staff"]
                state = format_pydantic_observation(data["data"], as_dict=True)
                # Add id information back into staff
                state["staff"]["staff_list"] = staff_list
                self.buffer.enqueue({data["state_type"]: state})
            elif data["state_type"] == "mid_day" and self.accept_midday_updates:
                self.num_updates += 1 
                data["data"]["staff"] = {"staff_list": data["data"]["staff"]}
                self.buffer.enqueue({'mid_day': data["data"]})
            elif data["state_type"] == "exit_time" and self.accept_midday_updates:
                self.buffer.enqueue({'exit_time': data["data"]})

        thread = threading.Thread(target=run)
        thread.daemon = True
        thread.start()

    def set_accept_midday_updates(self, accept: bool):
        """Enable/disable midday updates"""
        self.accept_midday_updates = accept

    def get_num_updates(self):
        """Get the number of updates received"""
        return self.num_updates

    def disconnect(self):
        """Disconnect from the server"""
        if self.sio:
            self.sio.disconnect()

    def is_connected(self):
        """Check if connected to server"""
        return self.sio and self.sio.connected

    def emit_action(self, action):
        """Emit an action to the server"""
        if self.sio and self.sio.connected:
            self.sio.emit('action', action)
        else:
            print("[SocketIO] Not connected to server, cannot emit action")


def create_game_state_listener():
    """Factory function to create a GameStateListener with a new queue"""
    buffer = Queue()
    return GameStateListener(buffer)

File: website/src/lib/map_js/gui/GameStateListener.js
import { io } from 'socket.io-client'

/**
 * Queue class equivalent to Python deque-based queue
 */
export class Queue {
  constructor() {
    this._items = []
  }

  enqueue(item) {
    this._items.push(item)
  }

  dequeue() {
    if (this.is_empty()) {
      throw new Error("dequeue from an empty queue")
    }
    return this._items.shift()
  }

  peek() {
    if (this.is_empty()) {
      return {}
    }
    return this._items[0]
  }

  is_empty() {
    return this._items.length === 0
  }

  size() {
    return this._items.length
  }

  clear() {
    this._items = []
  }

  toString() {
    return `Queue(${JSON.stringify(this._items)})`
  }
}

/**
 * GameStateListener class - JavaScript equivalent of Python GameStateListener
 */
export class GameStateListener {
  constructor(buffer, parkId, accept_midday_updates = false, callbacks = {}) {
    this.sio = null
    this.accept_midday_updates = accept_midday_updates

    // Step and buffer management
    this.buffer = buffer
    this.num_updates = 0
    this.parkId = parkId

    // Callbacks for GUI notifications
    this.onDisconnect = callbacks.onDisconnect || (() => {})
    this.onReconnect = callbacks.onReconnect || (() => {})
    this.onReconnectAttempt = callbacks.onReconnectAttempt || (() => {})
    this.onReconnectError = callbacks.onReconnectError || (() => {})
    this.onReconnectFailed = callbacks.onReconnectFailed || (() => {})

    // Track if we were previously connected (for reconnection detection)
    this.wasConnected = false
  }

  startSocketioListener(server = null) {
    // Use environment variable if server not provided
    if (!server) {
      // Detect environment based on hostname
      // If localhost, use localhost:3000, otherwise use relative path (empty string)
      const isLocal = typeof window !== 'undefined' &&
                     (window.location.hostname === 'localhost' || window.location.hostname === '127.0.0.1');
      server = isLocal ? 'http://localhost:3000' : '';
    }
    return new Promise((resolve, reject) => {
      try {
        // Development-only test mode: Use shorter timeouts to trigger disconnections faster
        const isDevelopment = server.includes('localhost') || server.includes('127.0.0.1');
        const testMode = false; // DEBUG: Set to true to test reconnection with aggressive 5s/3s timeouts

        this.sio = io(server, {
          transports: ['websocket', 'polling'],
          reconnection: true,
          reconnectionAttempts: 25,  // ~5 minutes with backoff
          reconnectionDelay: 1000,   // Start at 1 second
          reconnectionDelayMax: 10000,  // Max 10 seconds between attempts
          timeout: 10000,  // 10 second connection timeout (increased from default 20s)
          // In test mode, use very short intervals to trigger disconnection quickly
          pingInterval: (isDevelopment && testMode) ? 5000 : 60000,  // 5s (test) or 60s (prod)
          pingTimeout: (isDevelopment && testMode) ? 3000 : 30000    // 3s (test) or 30s (prod)
        })

        // Connection event handlers
        this.sio.on('connect', () => {
          console.log('[SocketIO] Connected to server')

          // If this is a reconnection (not initial connection), handle state recovery
          if (this.wasConnected) {
            console.log('[SocketIO] Reconnected after disconnection')
            this.onReconnect()
          }

          this.wasConnected = true
          resolve()
        })

        this.sio.on('disconnect', (reason) => {
          console.warn('[SocketIO] Disconnected:', reason)
          this.onDisconnect(reason)

          // If server initiated disconnect, manually reconnect
          if (reason === 'io server disconnect') {
            this.sio.connect()
          }
        })

        this.sio.on('connect_error', (error) => {
          // Only reject on initial connection error
          // After that, reconnection errors are handled separately
          if (!this.wasConnected) {
            reject(error)
          }
        })

        // Reconnection event handlers
        this.sio.on('reconnect_attempt', (attemptNumber) => {
          console.log(`[SocketIO] Reconnection attempt ${attemptNumber}`)
          this.onReconnectAttempt(attemptNumber)
        })

        this.sio.on('reconnect_error', (error) => {
          console.error('[SocketIO] Reconnection error:', error)
          this.onReconnectError(error)
        })

        this.sio.on('reconnect_failed', () => {
          console.error('[SocketIO] Reconnection failed after maximum attempts')
          this.onReconnectFailed()
        })

        // Add timeout
        setTimeout(() => {
          if (!this.sio.connected) {
            reject(new Error("Connection timeout"));
          }
        }, 5000);

        // Game event handlers
        this.sio.on('action', (data) => {
          console.log("[SocketIO] Action received:", data)
        })

        this.sio.on('game_update', (data) => {
          this.handleGameUpdate(data)
        })

      } catch (error) {
        console.error("[SocketIO] Failed to initialize:", error)
        reject(error)
      }
    })
  }

  handleGameUpdate(data) {
    const stateType = data.state_type;
    if (data.data.state.parkId !== this.parkId) {
      // console.log("Park ID mismatch:", data.data.state.parkId, "!=", this.parkId)
      return;
    }
    if (['full_state', 'day_start', 'day_end'].includes(stateType)) {
      const staffList = data.data.staff
      data.data.staff = {'staff_list': staffList}
      this.buffer.enqueue({ [stateType]: data.data })
      
    } else if (stateType === 'mid_day' && this.accept_midday_updates) {
      this.num_updates += 1
      data.data.staff = {'staff_list': data.data.staff}
      this.buffer.enqueue({ 'mid_day': data.data })
      
    } else if (stateType === 'exit_time' && this.accept_midday_updates) {
      this.buffer.enqueue({ 'exit_time': data.data })
    }
  }

  setAcceptMiddayUpdates(accept) {
    this.accept_midday_updates = accept
  }

  getNumUpdates() {
    return this.num_updates
  }

  disconnect() {
    if (this.sio) {
      this.sio.disconnect()
      this.sio = null
    }
  }

  isConnected() {
    return this.sio && this.sio.connected
  }

  emitAction(action) {
    if (this.sio && this.sio.connected) {
      this.sio.emit('action', action)
    } else {
      console.log("[SocketIO] Not connected to server, cannot emit action")
    }
  }
}

export function createGameStateListener(parkId) {
  const buffer = new Queue()
  return new GameStateListener(buffer, parkId)
}

// Export for use in other modules
export default {
  Queue,
  GameStateListener,
  createGameStateListener
}